[Goh Siau Chiak] iP

sc-arecrow opened the pull request.

Commit:
Add Gradle support

sc-arecrow committed:
Changed name from Duke to Viscount and changed logo

sc-arecrow committed:
Level-1

sc-arecrow committed:
Level-2

sc-arecrow committed:
Level-3

sc-arecrow committed:
Level-4

sc-arecrow committed:
A-TextUiTesting

sc-arecrow committed:
Added exceptions to handle errors

sc-arecrow committed:
Added delete function and changed error messages

sc-arecrow committed:
Refactor variable and method names and organise code

sc-arecrow committed:
Add javadoc to Viscount

sc-arecrow committed:
Modify constructor of Task and subclasses

Constructor of Task and subclasses Todo, Deadline and Event accept
one argument for its description, and the isDone field is set to
false.

To load task list data from disk, we must be able to create tasks
with isDone field set to true.

Let's add an isDone parameter to the constructor of Task and its
subclasses, which will allow us to dictate the isDone field when
we create new task objects.

sc-arecrow committed:
Make Task an abstract class and add toTaskData method

Task is a normal class.

There are some methods like toTaskData that are implemented
differently depending on the subclass of Task.

Let's make Task an abstract class to provide this common
functionality across the subclasses while also allowing Task to
have default methods.

Using an abstract class is preferable over an interface because
the subclasses have an "is-a" relationship to Task, instead of a
"can-do" relationship.

sc-arecrow committed:
Add save functionality

Viscount does not have save functionality.

Hence data is lost after the program ends.

Let's enhance Viscount with saving functionality so that tasks are
written onto disk after every change to the task list. Let's also
update the UI testing to simulate running Viscount multiple times
and test for persistence of data.

sc-arecrow committed:
Modify IOException handling

Viscount continues to run even after an IOException is thrown.

This runs the risk of users continuing to use Viscount when their
data is not being saved to disk.

Let's modify IOException handling such that when one is thrown,
Viscount immediately shuts down, preventing further loss of data.

sc-arecrow committed:
Improve date and time representation

Viscount represents date and time with a string given by the user.

This is inadequate because of the lack of standardisation of date
and time across different tasks.

Let's improve date and time representation using the java
LocalDateTime library so it is standardised across all tasks.

sc-arecrow committed:
Make Task an abstract class and add hasDateTime method

Task is a normal class.

There are some methods like hasDateTime that are implemented
differently depending on the subclass of Task.

Let's make Task an abstract class to provide this common
functionality across the subclasses while also allowing Task to
have default methods.

Using an abstract class is preferable over an interface because the
subclasses have an "is-a" relationship to Task, instead of a
"can-do" relationship.

sc-arecrow committed:
Create TaskType enum

There is no TaskType enum.

Objects that are instances of subclasses of Task cannot be
differentiated according to their type of task.

Let's create a TaskType enum and give each instance of Task a
taskType field to allow for this differentiation.

sc-arecrow committed:
Improve list function

Viscount can list tasks that users have added.

With a better representation of date and time of deadlines and events,
users should be able to have better control over what tasks are listed.

Let's improve the list function to enable users to:
* list tasks of a specified type
* list tasks occurring today or a specified day

sc-arecrow committed:
Merge branch 'branch-Level-7'

sc-arecrow committed:
Merge branch 'branch-Level-8' and fix bug with delete save data

Deleting tasks may result in corrupted save data file due to
incorrect StandardOpenOption given to write method, resulting in
direct writing to data file without deleting its previous contents
first.

Let's fix the option given by removing it so that the file is
truncated to length 0 first, then overwritten.

sc-arecrow committed:
Extracted Ui, Storage, Parser, TaskList and Command classes.

Viscount class is handling many functionalities, of which most
can be classified into various categories.

No one class should hold that many functions as it goes against the
Single Responsibility Principle in OOP.

Let's extract out code with similar functions into several classes,
namely Ui, Storage, Parser, TaskList and Command, to better represent
the system in an object oriented manner.

sc-arecrow committed:
Added several exception handling cases

Viscount does not have exception handling for IO exceptions and
using the following command 'list todo /on'.

Since these are exceptions that can reasonably occur, they are checked
exceptions and should be handled accordingly.

Let's:
* Add a new exception for unsupported operations
* Gracefully handle IOExceptions by shutting down Viscount when they
occur, which will potentially save the user from the loss of their data

sc-arecrow committed:
Divide classes into packages

Viscount code is all in one directory.

Class files that are related to one another can be grouped into
packages. This results in a more organised project structure.

Let's organise the class files into suitable java packages.

sc-arecrow committed:
Add JUnit test for some Ui and TaskList functions

Viscount is not undergoing any unit testing.

It is difficult to confirm that Viscount is working as intended with
only text UI testing. It is also beneficial to test each part of the
program separately.

Let's add JUnit tests for some of the functions in Ui and TaskList
to confirm that these parts of Viscount are working as intended.

sc-arecrow committed:
Fix bug for Viscount to create data directory

Viscount shuts down if it cannot find the data directory in which
the data file is supposed to be in.

Viscount should create the directory if it cannot find it, before
creating the data file inside.

Let's fix the bug to enable Viscount to create the data directory.

sc-arecrow committed:
Fix runtest.bat to write correct Unicode symbols

When runtest.bat is run, the output written to ACTUAL.TXT contains
? characters instead of the correct tick and cross Unicode characters.

This is an inaccurate representation of the true output from Viscount.

Let's fix the runtest.bat to write the correct Unicode symbols by
encoding with UTF-8.

sc-arecrow committed:
Fix runtest.bat to write correct Unicode symbols

When runtest.bat is run, the output written to ACTUAL.TXT contains
? characters instead of the correct tick and cross Unicode characters.

This is an inaccurate representation of the true output from Viscount.

Let's fix the runtest.bat to write the correct Unicode symbols by
encoding with UTF-8.

sc-arecrow committed:
Merge branch 'master' of https://github.com/sc-arecrow/viscount

sc-arecrow committed:
Merge branch 'master' into branch-A-JavaDoc

sc-arecrow committed:
Refactor Parser class and change add task command

Parser class contains parse method that contains all logic pertaining
to parsing of raw input from user. User types task type followed by
arguments to add a new task.

This is not as obvious to the user, who may accidentally add tasks
that they did not intend to.

Let's
* extract out the parts of code from parse method to make it more
readable
* change add task command so users have to type 'add' followed by
the task type and then the arguments

sc-arecrow committed:
Add JavaDocs

Viscount code does not contain any Javadocs

Javadocs are useful for other developers to quickly understand
what a method does, and hence should be added for public methods
that are accessible to all developers.

Let's add Javadoc comments to Viscount code.

sc-arecrow committed:
Tweak code to comply with coding standard.

Viscount code has some inconsistencies with coding standard.

Code should comply to a coding standard for readability.

Let's tweak Viscount code to comply with a coding standard.

Refer to this page to see the coding standards
https://se-education.org/guides/conventions/java/intermediate.html

sc-arecrow committed:
Add Javadocs to Viscount classes

Viscount classes do not have any header comments.

Viscount classes should have header comments to let other developers
know what each class represents, especially for public classes.

Let's add Javadocs to Viscount classes.

sc-arecrow committed:
Rename modifier to taskTypeModifier for list command

List command contains a variable called 'modifier'.

This is an ambiguous name that does not fully represent what
information the variable is supposed to contain.

Let's rename the variable to taskTypeModifier.

sc-arecrow committed:
Add find functionality to list command

Users can list their tasks and can filter by task type or by date
and time of task.

Users should also be able to filter their tasks by the description.

Let's add a find functionaly to the list command to allow users to
filter by description in conjunction with the rest of the list
modifiers.

sc-arecrow committed:
Merge branch 'branch-A-JavaDoc'

sc-arecrow committed:
Merge branch 'branch-A-CodingStandard'

sc-arecrow committed:
Merge branch 'branch-Level-9'

sc-arecrow committed:
Add more tests to UiTest.

UiTest tests showWelcome, showExit and input.

UiTest should test for the other functions of the Ui class, especially
those listing and modifying tasks and the task list.

Let's add more tests to UiTest to test listing and modifying tasks and
the task list.

sc-arecrow committed:
Merge remote-tracking branch 'origin/add-gradle-support'

sc-arecrow committed:
Add Gradle support

Viscount does not have any build automation tools implemented.

Build automation tools like Gradle can help the process of building
the project by handling various aspects like testing and dependency
management.

Let's add Gradle support to Viscount.

sc-arecrow committed:
Add CheckStyle to the project.

Coding style for Viscount is checked manually.

This may result in oversight of some coding style violations due to
human error.

Let's add CheckStyle, a static analysis tool, to help detect coding
style violations in the project automatically.

sc-arecrow committed:
Setup basic JavaFX in Viscount

Viscount uses the console to interact with the user.

A Graphical User Interface (GUI) gives the user a better experience
with using Viscount because of easier actions like mouse clicks, and
also because of visual feedback to their actions.

Let's setup JavaFX in Viscount.

sc-arecrow committed:
Implement basic GUI for chat function.

With JavaFX set up, the next step is to make the screen for the chat
window.

Let's create a basic GUI for the chat window, to be integrated with
Viscount later.

sc-arecrow committed:
Convert basic GUI to use FXML

The basic GUI setup has many major problems:
* Every small change to the GUI requires rebuilding the app
* Components are heavily dependent on each other
* Code is untidy and long

Let's convert the basic GUI to use FXML instead.

kwannoel commented on src/main/java/viscount/Storage.java:
```diff
@@ -0,0 +1,120 @@
+                            tasks.add(new Todo(taskDescription, isDone));
+                        } else if (taskType == TaskType.Deadline) {
+                            LocalDateTime dueDate = Parser.parseDateTime(
+                                    taskData.get(3), Parser.TASK_DATA_DATE_TIME_FORMATTER);
```
```suggestion
                                    taskData.getDateTime(), Parser.TASK_DATA_DATE_TIME_FORMATTER);
```

sc-arecrow replied:
Since `taskData` refers to the `List<String>` split from the data saved in the text file, it does not have a `getDateTime` method. The reason for calling the `get` method on the `List<String>` with the input `3` is to get the date time string.

PrestonTYR commented on src/main/java/viscount/command/AddCommand.java:
```diff
@@ -0,0 +1,65 @@
+
+import viscount.*;
+
+import viscount.task.*;
```
Hey, minor nitpick, but I believe that according to the coding standard, imports should all be listed explicitly. I noticed the same issue in the other classes as well

sc-arecrow replied:
Thanks for the review! I've made the changes according to coding standard :)

sc-arecrow committed:
Integrate Viscount logic with JavaFX GUI and FXML

After creating a GUI for the chat window on JavaFX, the next step is
to integrate Viscount chatbot logic with the GUI.

Let's restructure the Viscount logic to change it from a console-based
application to one with a GUI, implemented with JavaFX and FXML

sc-arecrow committed:
Merge branch 'branch-A-Gradle'

sc-arecrow committed:
Merge branch 'branch-A-CheckStyle'

sc-arecrow committed:
Merge branch 'branch-Level-10'

PrestonTYR reviewed (commented):
Well formatted code that adheres to the coding standards of the mod, with very minor mistakes. I could only find 1 area where the coding standard isn't followed after reading through the code

sc-arecrow committed:
Fix bug where application could not be launched from JAR file

Viscount application works when the main method from the Main class
was ran.

However, trying to launch the application using the JAR file resulted
in an error.

Let's fix the bug to allow users to launch the application from the JAR
file.

sc-arecrow committed:
Add assertions to ensure code logic holds.

Viscount code does not contain any assertions.

Assertions should be implemented in any program to verify the program
state, and an assertion failure indicates a possible bug in the code.

Let's add assertions to verify Viscount code logic and to look out for
possible bugs in said logic.

sc-arecrow committed:
Refactored Viscount code to improve code quality.

Viscount code contains methods that are very long and violates the
SLAP principle in some places.

Understandability of code is important as code may have to be read,
understood and modified by other developers later on.

Let's improve code quality in Viscount to make it more readable.

sc-arecrow committed:
Merge pull request #1 from sc-arecrow/branch-A-Assertions

Add assertions to ensure code logic holds.

sc-arecrow committed:
Merge branch 'master' into branch-A-CodeQuality

sc-arecrow committed:
Merge pull request #2 from sc-arecrow/branch-A-CodeQuality

Refactored Viscount code to improve code quality.

sc-arecrow committed:
Setup Continuous Integration for Viscount

Viscount does not have any Continuous Integration put in place.

Continuous Integration is a form of build automation that allows developers to integrate code changes continuously by running automated processes usually after a push or merge into master.

Let's setup simple Continuous Integration for Viscount.

gsmoon97 commented on src/main/java/viscount/Storage.java:
```diff
@@ -0,0 +1,125 @@
+     *
+     * @param dataDirectoryPath Directory path to the data file.
+     */
+    public Storage(String dataDirectoryPath) {
+        this.dataDirectoryPath = dataDirectoryPath;
+        this.filePathString = dataDirectoryPath + DATA_FILE_NAME;
+    }
```
Do you think it might be a good idea to specify the data directory path inside this class as constant instead of taking it in as an argument because the data directory is fixed anyway?

gsmoon97 commented on src/main/java/viscount/Viscount.java:
```diff
@@ -0,0 +1,52 @@
+
+
+/**
+ * Represents Viscount, a chatbot that helps the user keep track of tasks.
```
Not an issue, but a question. What does Viscount stand for? :)

gsmoon97 commented on src/main/java/viscount/command/ListCommand.java:
```diff
@@ -0,0 +1,97 @@
+
+        List<Task> tasks = taskList.getTasks();
+
+        if (dateString.isEmpty()) {
+            List<Task> filteredTasks = tasks
+                    .stream()
+                    .filter(filterByModifier)
+                    .filter(filterByDescription)
+                    .collect(Collectors.toList());
```
This is a very nice use of stream here. Keep up the good work 👍

gsmoon97 commented on src/main/java/viscount/gui/MainWindow.java:
```diff
@@ -0,0 +1,81 @@
+     */
+    public MainWindow() {
+        try {
+            this.viscount = new Viscount(DATA_DIRECTORY_PATH);
+            FXMLLoader fxmlLoader = new FXMLLoader(Main.class.getResource("/view/MainWindow.fxml"));
+            fxmlLoader.setController(this);
+            fxmlLoader.setRoot(this);
+            fxmlLoader.load();
```
I realized this part is quite different from the example. Could you briefly explain why you created a new fxmlLoader object inside the MainWindow class instead of the Main class?

gsmoon97 reviewed (commented):
Overall, you have done a superb job so far. You have followed the coding standards perfectly and also have organized the classes into appropriate packages, so it was a pleasure for me to review your code. I especially liked how you incorporated various Java components such as Stream to make your code more efficient. Keep up the good work 👍

sc-arecrow committed:
Add mass done and delete operations

Viscount users can only mark done and delete tasks one at a time, by
using their index number in the list.

This is not user friendly especially if the user has many tasks that
they want to mark as done or delete all at once.

Let's add
* a mark all tasks as done operation
* a delete all tasks operation
* a delete all done tasks operation

sc-arecrow committed:
Remove redundant exceptions and clean up exception handling

These are some exceptions in Viscount:
* ViscountMissingArgumentException
* ViscountMissingDescriptionException
* ViscountMissingArgumentDescriptionException

These exceptions are similar to one another and it is redundant to
have multiple exception classes.

Let's:
* remove the redundant exceptions and combine their functions to
ViscountMissingArgumentException
* clean up some parts of exception handling

sc-arecrow committed:
Override equals method to subclasses of Task

Todo, Deadline and Event does not override equals method.

This results in task comparison to fail when the same task is
instantiated twice into two objects having equal fields, because the
default equals method compares the objects' hashcode.

Let's override the equals method to declare that tasks are equal if
they have exactly the same fields (task type, description, doneness
and date time if applicable).

sc-arecrow committed:
Override equals method of subclasses of Command

Add, Delete, Done and ListCommands do not have an equals method
overriding the default method, which compares objects by reference.

This results in command comparison failure when comparing the same
command instantiated into two objects with the same fields.

Let's override the equals method to declare command equality when
the two commands have the same fields.

sc-arecrow committed:
Add JUnit tests for Parser class

Viscount has JUnit tests for its Ui and TaskList classes.

Let's enhance Viscount testing by adding tests for the Parser class,
to ensure the parser is translating user input correctly.

sc-arecrow committed:
Improve Viscount GUI

Viscount has basic default GUI from JavaFX.

A more visually appealing GUI might make the application more
enjoyable to use.

Let's improve Viscount's GUI by
* introducing some padding to make messages spaced out properly
* putting input and responses in visible text boxes
* setting error responses to have a different background color
* setting an application icon and title

sc-arecrow committed:
Change application name to Viscount

Application name is "Viscount Task Manager"

Let's change the name to "Viscount"

sc-arecrow committed:
Add User Guide to Viscount

Prospective users of Viscount do not have any guides to refer to to
figure out how to use Viscount.

Let's add a User Guide so that new users are able to learn the
various features that Viscount has.

sc-arecrow committed:
Add edit feature into Viscount

Viscount does not have an edit function.

Users have to delete and recreate a task from scratch if they make
a mistake while adding a task.

Let's add an edit task feature so users can edit existing tasks with
less hassle.

sc-arecrow committed:
Add JUnit tests for edit feature

Viscount has edit feature where users can update their tasks without
deleting and remaking a new task.

Let's add JUnit tests for the edit feature.

sc-arecrow committed:
Update README.md

README.md contains Duke default instructions.

Let's update the README.md to tell both users and
developers how to set up Viscount.

sc-arecrow committed:
Update README.md

Fixed a typo in the user guide.

sc-arecrow committed:
Update README.md

Fixed a typo in the README.md

sc-arecrow committed:
Add Javadocs to public methods

Some public methods in Viscount code do not have Javadoc comments.

According to the coding standard, all public classes and methods
must have Javadoc comments.

Let's add Javadocs to these public methods that are missing them.

sc-arecrow committed:
Merge branch 'master' of https://github.com/sc-arecrow/viscount

sc-arecrow committed:
Set theme jekyll-theme-slate

sc-arecrow committed:
Update README.md

Changed the link of the user guide

sc-arecrow committed:
Update README.md

Typo in `delete done` usage
